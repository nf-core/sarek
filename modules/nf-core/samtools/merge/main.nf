process SAMTOOLS_MERGE {
    tag "$meta.id"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/samtools:1.21--h50ea8bc_0' :
        'biocontainers/samtools:1.21--h50ea8bc_0' }"

    input:
    tuple val(meta), path(input_files, stageAs: "?/*")
    tuple val(meta2), path(fasta)
    tuple val(meta3), path(fai)

    output:
    tuple val(meta), path("${prefix}.bam") , optional:true, emit: bam
    tuple val(meta), path("${prefix}.cram"), optional:true, emit: cram
    tuple val(meta), path("*.csi")         , optional:true, emit: csi
    tuple val(meta), path("*.crai")        , optional:true, emit: crai
    path  "versions.yml"                                  , emit: versions


    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args   ?: ''
    prefix   = task.ext.prefix ?: "${meta.id}"
    def file_type = input_files instanceof List ? input_files[0].getExtension() : input_files.getExtension()
    def reference = fasta ? "--reference ${fasta}" : ""
    """
    samtools \\
        merge \\
        --threads ${task.cpus-1} \\
        $args \\
        ${reference} \\
        ${prefix}.${file_type} \\
        $input_files

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args   ?: ''
    prefix = task.ext.suffix ? "${meta.id}${task.ext.suffix}" : "${meta.id}"
    def file_type = input_files instanceof List ? input_files[0].getExtension() : input_files.getExtension()
    def index_type = file_type == "bam" ? "csi" : "crai"
    def index = args.contains("--write-index") ? "touch ${prefix}.${index_type}" : ""
    """
    touch ${prefix}.${file_type}
    ${index}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """
}
