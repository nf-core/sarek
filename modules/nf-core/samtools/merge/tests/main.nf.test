nextflow_process {

    name "Test Process SAMTOOLS_MERGE"
    script "../main.nf"
    process "SAMTOOLS_MERGE"

    tag "modules"
    tag "modules_nfcore"
    tag "samtools"
    tag "samtools/merge"

    test("bams") {

        config "./index.config"

        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.methylated.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.single_end.sorted.bam', checkIfExists: true) ]
                ])
                input[1] = [[],[]]
                input[2] = [[],[]]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(file(process.out.bam[0][1]).name).match("bams_bam") },
                { assert snapshot(process.out.cram).match("bams_cram") },
                { assert snapshot(file(process.out.csi[0][1]).name).match("bams_csi") },
                { assert snapshot(process.out.crai).match("bams_crai") },
                { assert snapshot(process.out.versions).match("bams_versions") }
            )
        }
    }

    test("crams") {

        config "./index.config"

        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/cram/test.paired_end.recalibrated.sorted.cram', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/cram/test2.paired_end.recalibrated.sorted.cram', checkIfExists: true) ]
                ])
                input[1] = Channel.of([
                    [ id:'genome' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                input[2] = Channel.of([
                    [ id:'genome' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(file(process.out.cram[0][1]).name).match("crams_cram") },
                { assert snapshot(process.out.bam).match("crams_bam") },
                { assert snapshot(file(process.out.crai[0][1]).name).match("crams_crai") },
                { assert snapshot(process.out.csi).match("crams_csi") },
                { assert snapshot(process.out.versions).match("crams_versions") }
            )
        }
    }

    test("bam") {

        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.methylated.sorted.bam', checkIfExists: true) ]
                ])
                input[1] = [[],[]]
                input[2] = [[],[]]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(file(process.out.bam[0][1]).name).match("bam_bam") },
                { assert snapshot(process.out.cram).match("bam_cram") },
                { assert snapshot(process.out.crai).match("bam_crai") },
                { assert snapshot(process.out.csi).match("bam_csi") },
                { assert snapshot(process.out.versions).match("bam_versions") }
            )
        }
    }

    test("bams_stub") {

        config "./index.config"
        options "-stub"

        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.methylated.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.single_end.sorted.bam', checkIfExists: true) ]
                ])
                input[1] = [[],[]]
                input[2] = [[],[]]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(file(process.out.bam[0][1]).name).match("bams_stub_bam") },
                { assert snapshot(process.out.cram).match("bams_stub_cram") },
                { assert snapshot(file(process.out.csi[0][1]).name).match("bams_stub_csi") },
                { assert snapshot(process.out.crai).match("bams_stub_crai") },
                { assert snapshot(process.out.versions).match("bams_stub_versions") }
            )
        }
    }
}
