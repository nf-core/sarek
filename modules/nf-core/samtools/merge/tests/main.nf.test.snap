{
    "crams_cram": {
        "content": [
            "test.cram"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:50:00.647389"
    },
    "bams_stub_cram": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:50:19.937013"
    },
    "bams_crai": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:49:24.928616"
    },
    "bams_bam": {
        "content": [
            "test.bam"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:49:24.923289"
    },
    "bams_cram": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:49:24.925716"
    },
    "crams_csi": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:50:00.655959"
    },
    "bam_bam": {
        "content": [
            "test.bam"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:50:10.319539"
    },
    "bam_versions": {
        "content": [
            [
                "versions.yml:md5,d51d18a97513e370e43f0c891c51dfc4"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-16T09:16:30.476887194"
    },
    "bams_csi": {
        "content": [
            "test.bam.csi"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:49:24.92719"
    },
    "bams_stub_csi": {
        "content": [
            "test.csi"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:50:19.940498"
    },
    "bam_crai": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:50:10.328852"
    },
    "bams_stub_versions": {
        "content": [
            [
                "versions.yml:md5,d51d18a97513e370e43f0c891c51dfc4"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-16T09:16:52.203823961"
    },
    "bam_cram": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:50:10.324219"
    },
    "bams_stub_bam": {
        "content": [
            "test.bam"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:50:19.933153"
    },
    "bams_versions": {
        "content": [
            [
                "versions.yml:md5,d51d18a97513e370e43f0c891c51dfc4"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-16T08:29:57.524363148"
    },
    "crams_bam": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:50:00.650652"
    },
    "crams_versions": {
        "content": [
            [
                "versions.yml:md5,d51d18a97513e370e43f0c891c51dfc4"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-16T09:16:06.977096207"
    },
    "bam_csi": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:50:10.33292"
    },
    "crams_crai": {
        "content": [
            "test.cram.crai"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:50:00.653512"
    },
    "bams_stub_crai": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-02-12T18:50:19.943839"
    }
}